core/src/main/java/ai/timefold/solver/core/impl/localsearch/decider/reconfiguration/RestoreBestSolutionReconfigurationStrategy.java

```diff
@@ -0,0 +1,73 @@
+package ai.timefold.solver.core.impl.localsearch.decider.reconfiguration;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.impl.heuristic.selector.move.MoveSelector;
+import ai.timefold.solver.core.impl.localsearch.decider.acceptor.Acceptor;
+import ai.timefold.solver.core.impl.localsearch.scope.LocalSearchPhaseScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RestoreBestSolutionReconfigurationStrategy<Solution_> implements ReconfigurationStrategy<Solution_> {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final MoveSelector<Solution_> moveSelector;
+    private final Acceptor<Solution_> acceptor;
+
+    public RestoreBestSolutionReconfigurationStrategy(MoveSelector<Solution_> moveSelector, Acceptor<Solution_> acceptor) {
+        this.moveSelector = moveSelector;
+        this.acceptor = acceptor;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <Score_ extends Score<Score_>> Score_ apply(AbstractStepScope<Solution_> stepScope) {
+        var solverScope = stepScope.getPhaseScope().getSolverScope();
+        logger.debug("Resetting working solution, score ({})", solverScope.getBestScore());
+        solverScope.setWorkingSolutionFromBestSolution();
+        // Changing the working solution requires reinitializing the move selector and acceptor
+        // 1 - The move selector will reset all cached lists using old solution entity references
+        moveSelector.phaseStarted(stepScope.getPhaseScope());
+        // 2 - The acceptor will restart its search from the updated working solution (last best solution)
+        acceptor.phaseStarted((LocalSearchPhaseScope<Solution_>) stepScope.getPhaseScope());
```

Comment:
Looks like this code should be shared with something somewhere else. (Sorry for not being more specific, I don't remember 100 % of our codebase.)

If something _has to happen_ when we are restoring the working solution, then arguably every piece of code that restores the working solution should share this code. This is how you prevent bugs.

Reply:
The logic has been moved to `LocalSearchDecider`.